### PR TITLE
545: Missing HttpServer auto gzip decompression in request handling

### LINKED ISSUE #545: Missing HttpServer auto gzip decompression in request handling
**Citrus Version**
v2.7.8

**Expected behavior**
Auto handle gzip compressed requests on Http server component with automatic decompression of request payload.

**Actual behavior**
Server skips decompression and performs default toString conversion of binary zipped input stream which leads to irreversible loss/corruption of payload data.

**Test case sample**
```xml
      <echo>
        <message>Send Http client request with gzip compressed data</message>
      </echo>

      <http:send-request client="httpClient" fork="true">
        <http:POST>
          <http:headers content-type="text/html">
            <http:header name="Content-Encoding" value="gzip"/>
          </http:headers>
          <http:body type="gzip">
            <http:data>Hello</http:data>
          </http:body>
        </http:POST>
      </http:send-request>

      <echo>
        <message>Receive gzip decompressed data on server side</message>
      </echo>

      <http:receive-request server="httpServer">
        <http:POST path="/echo">
          <http:headers>
            <http:header name="Content-Encoding" value="gzip"/>
          </http:headers>
          <http:body type="plaintext">
            <http:data>Hello</http:data>
          </http:body>
        </http:POST>
      </http:receive-request>
```

### PR DESCRIPTION
**Pull request by christophd**
Fixes #545

### CODE REVIEW

File: modules/citrus-http/src/test/resources/com/consol/citrus/http/integration/HttpGzipEncodedIT.xml [hunk 1]
```diff
@@ -85,7 +85,7 @@
           <http:headers content-type="text/html">
             <http:header name="Content-Encoding" value="gzip"/>
           </http:headers>
-          <http:body>
+          <http:body type="gzip">
```
Comment (svettwer) [inline modules/citrus-http/src/test/resources/com/consol/citrus/http/integration/HttpGzipEncodedIT.xml]: I think the `type` declaration is something we should get rid of n the long term. I see that this is not an option as part of this change but we should consider this for a future major release. And yes, I see that this is a shitload of work. :sweat_smile:

Comment (svettwer) [reply modules/citrus-http/src/test/resources/com/consol/citrus/http/integration/HttpGzipEncodedIT.xml]: Now as I think about it myself, I don't know whether this is possible at all. It's important to keep in mind, that we do not only handle plaintext payload. If we've resource files, we'd have to specify the content type of the resource file anyways as we don't know if it's a plain text content or a gzip resource with prepared test data.

**Comment by svettwer** [comment]
Happy new year!

Do you have any ideas here @christophd ?

**Comment by christophd** [comment]
Added fix for the Http Java DSL client side compression

**Comment by christophd** [comment]
Build is unstable because of some other test I did not touch `JdbcConnectionIT` Just doing a rebuild

**Comment by christophd** [comment]
We really have to do something about the fragile CI build. 2nd run ended in another unrelated test failing.

**Comment by svettwer** [comment]
> We really have to do something about the fragile CI build. 2nd run ended in another unrelated test failing.

The mere fact that we now have automatic branch builds with attached static code analysis is a huge advantage. Yes, the build is unstable and that should be improved. However, due to the grown structures, probably larger changes are necessary. Unfortunately, this is nothing I can take care about in the near future. I think this 80% solution is okay-isch in the first place.

**Comment by christophd** [comment]
SonarCloud Code Analysis check seems to be hanging for quite some time. @svettwer any idea?

**Comment by svettwer** [comment]
It seems like the build was not transmitted to the sonarcloud server or not recognized by sonarcloud itself for some reason. As this branch is stale in comparison to v2.7-bugfix, a rebase including a rebuild is required anyways. This will trigger the sonarcloud analysis again.

**Comment by svettwer** [comment]
Hmhm that's strange. We haven't had this issue before.

**Comment by svettwer** [comment]
Got it!
Sonacloud said: `Due to an upgrade, you need first to re-analyze the target branch 'v2.7-bugfix' before analyzing this pull request.`

**Comment by svettwer** [comment]
I triggered a new analysis of v2.7-bufix. After this, the analysis should work as it should.

**Comment by christophd** [comment]
I am afraid that did not help much

**Comment by svettwer** [comment]
It helped but I forgot to rebuild your branch. 😅
Here is your analysis now. 👍😁

File: modules/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/SendMessageBuilder.java [hunk 2]
```diff
@@ -432,6 +439,15 @@ public T messageType(MessageType messageType) {
     public T messageType(String messageType) {
         this.messageType = messageType;
         getAction().setMessageType(messageType);
+
+        if (MessageType.BINARY.name().equals(messageType)) {
```
Comment (svettwer) [inline modules/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/SendMessageBuilder.java]: I think we should use `binaryMessageConstructionInterceptor.supportsMessageType()` here to have one source of truth for the decision, whether the interceptor could handle the message type.

File: modules/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/SendMessageBuilder.java [hunk 3]
```diff
@@ -432,6 +439,15 @@ public T messageType(MessageType messageType) {
...
+            getMessageContentBuilder().add(binaryMessageConstructionInterceptor);
+        }
+
+        if (MessageType.GZIP.name().equals(messageType)) {
```
Comment (svettwer) [inline modules/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/SendMessageBuilder.java]: I think we should use `gzipMessageConstructionInterceptor.supportsMessageType()` here to have one source of truth for the decision, whether the interceptor could handle the message type.

**Comment by svettwer** [comment]
Hi!

I know there are a lot of minor changes to do. If you can't manage to apply them, I'll do it during the next weeks, so that we can close this PR finally. =)
Thx for your efforts! 

BR,
Sven

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.zip.GZIPInputStream;
+
+import com.consol.citrus.exceptions.CitrusRuntimeException;
+
+/**
+ * Request wrapper wraps gzip input stream with unzipped stream. Read operations on that stream are
+ * automatically decompressed with gzip encoding.
+ *
+ * @author Christoph Deppisch
+ * @since 2.8.0
```
[the remark this answers is not in the document]
Comment (christophd) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Fixing this using the github online editor would have been much easier than writing this review comment

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+ */
+public class GzipHttpServletRequestWrapper extends HttpServletRequestWrapper {
+    /**
+     * Constructs a request adaptor wrapping the given request.
+     *
+     * @param request
```
[the remark this answers is not in the document]
Comment (christophd) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Really?

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            super();
+            gzipStream = new GZIPInputStream(request.getInputStream());
+        }
+
+        @Override
+        public boolean isFinished() {
+            try {
+                return gzipStream.available() == 0;
+            } catch (IOException e) {
+                throw new CitrusRuntimeException("Failed to check gzip intput stream availability", e);
```
[the remark this answers is not in the document]
Comment (christophd) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Fixing this using the github online editor would have been much easier than writing this review comment

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * Gzip enabled servlet input stream.
+     */
+    private class GzipServletInputStream extends ServletInputStream {
+        private final GZIPInputStream gzipStream;
+
+        /**
+         * Default constructor using wrapped input stream.
```
[the remark this answers is not in the document]
Comment (christophd) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Missing on that field?

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (svettwer) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Not an issue at all, just wanted to let you know.

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (svettwer) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Same here, not an issue at all, just wanted to let you know.

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (svettwer) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Maybe I got the wrong line number... :thinking: :man_shrugging:

File: modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (svettwer) [reply modules/citrus-http/src/main/java/com/consol/citrus/http/servlet/GzipHttpServletRequestWrapper.java]: Same here, not an issue at all, just wanted to let you know.

**Comment by svettwer** [comment]
Just a side note: We've discussed this. I'll take care of the changes.

**Pull request merged**